[PR] MSI: 619: MFTF Tests for MSI - Part 2

ivy00johns opened the pull request:
### Description
This is the second set of Test Cases created for MSI using MFTF. This PR contains 14 MSI specific Tests. Please reference the following Wiki for instructions on how to run them: https://github.com/magento-engcom/msi/wiki/Getting-Started

magento-engcom/msi#619 [MFTF] Automated MSI Test Cases - Part 2

### Known Skipped Tests
The following Tests are skipped due to the following Bug: https://app.zenhub.com/workspace/o/magento-engcom/msi/issues/815
* AdminMultipleWebsitesAssignedAsSalesChannelsToStockTest
* AdminStockCannotBeDeletedWhenSalesChannelIsAssignedTest
* AdminNewWebsiteHasDefaultStockAssignedAutomaticallyTest
* AdminCreateNewWebsiteWithStoreAndStoreViewTest

The following Test is skipped due to the following Bug: https://app.zenhub.com/workspace/o/magento-engcom/msi/issues/844
* AdminStockDeletedFromGridWhenNoSalesChannelsAssignedTest 

The following Test is skipped due to the following Bug: https://jira.corp.magento.com/browse/MQE-914
* AdminAddConfigurableProductWithDropDownAttributeAndDefaultSourceTest

You can search for the phrase "SKIPPING DUE TO BUG" to locate all of the Tests.

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Adding new MSI Tests.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Adding Add/Remove column Tests for Sources/Stock.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Merged 727 into this branch so I could use my Page Objects/Data/Metadata.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Adding Descriptions to 2 Tests.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Merging 782: MSI MFTF Test Cases 1

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Adding a new test.
- Adding waitForPageLoad after a few actions to resolve flaky test issues.

ivy00johns committed:
833: MSI Test Cases 2

- Merging in changes from MSI Test Cases 1

ivy00johns committed:
833: MSI Test Cases 2

- Removing old Stock/Metadata folders. The real versions are listed under the InventoryApi module.
- Updating a few Tests to correctly use the API.

ivy00johns committed:
Squashed commit of the following:

commit dceb49854bf7a04ad400e1f9fa1203a61eff78dc
Author: John Stennett <john00ivy@gmail.com>
Date:   Thu Mar 22 11:12:26 2018 -0500

    727: MSI Test Cases 1

    - Moving all Inventory Test files to /dev/tests/acceptance from /app/code/.
    - Adding composer.json, license files and README to the new folders.

commit 8d25712718cae30f65f327c3ca135b3b55ba385c
Author: John Stennett <john00ivy@gmail.com>
Date:   Thu Mar 22 10:57:39 2018 -0500

    727: MSI Test Cases 1

    - Updating selector for the Filter drop down menu and moved it to a section.
    - Updated Tests to use the section element instead of being hardcoded.

commit 7628980d49eb811613e01808748643ca8ffa697a
Author: John Stennett <john00ivy@gmail.com>
Date:   Wed Mar 21 15:52:02 2018 -0500

    782: MSI MFTF Test Cases 1

    - Removing an incomplete test.
    - Replacing the Admin Form Post metadata with an API metadata alternative.
    - Updating the Tests to use the API post instead of the Form Post.

commit 155e917b550ff5108f81d52d63d286ed6fd0c270
Author: John Stennett <john00ivy@gmail.com>
Date:   Tue Mar 20 13:46:14 2018 -0500

    782: MSI MFTF Test Cases 2

    - Removing data that is no longer needed.

commit d7b94e5f8bb3b034454e01ebcd5a8edd509d77db
Author: John Stennett <john00ivy@gmail.com>
Date:   Mon Mar 19 10:55:17 2018 -0500

    782: MSI MFTF Test Cases 2

    - Renamed "AdminSaveAndClose" action group to "AdminFormSaveAndClose".

commit b4f183c1bf846a06267a5b78ac332cd435609a4b
Author: John Stennett <john00ivy@gmail.com>
Date:   Tue Mar 13 16:03:45 2018 -0500

    782: MSI MFTF Test Cases 2

    - Addressed CR concerns.
    - Moved the API files to the correct module.
    - Added addition asserts that I missed.
    - Updated a URL to match the changes we made in MFTF.

commit 76e835e8d7b394d4228de04f37a91804a8bbfe84
Author: John Stennett <john00ivy@gmail.com>
Date:   Wed Feb 28 09:46:15 2018 -0600

    782: MSI MFTF Test Cases

    - Fixed a selector. It was missing a "."
    - Updated the Stock Data name.

commit c1921053faa891454d9a5d94bcd97dd80a65936c
Author: John Stennett <john00ivy@gmail.com>
Date:   Tue Feb 27 10:49:08 2018 -0600

    782: MSI MFTF Test Cases

    - Updating metadata URL attribute name to match newest schema changes.

commit 7da2582770b784c15bcf0f23b24ec88c86e1c286
Author: John Stennett <john00ivy@gmail.com>
Date:   Fri Feb 23 13:31:39 2018 -0600

    782: MSI MFTF Test Cases

    - Adding new attribute to the metadata so the URL is generated correctly for the Form Post.
    - Fixed an invalid element "type" for a few selectors.
    - Replacing old action group with a conditional click for the "Clear All" button.
    - Moved all Tests to the "acceptance" directory instead of "functional".

commit 050ffd2e2e9205a14dda93b397dee11986d7ef5f
Author: John Stennett <john00ivy@gmail.com>
Date:   Thu Feb 8 11:27:38 2018 -0600

    782: MST MFTF Test Cases

    - Replacing actions with the new Action Groups I setup.

commit 68a4b34d9cc2a63a51f073acd4b729e2cae78c71
Author: John Stennett <john00ivy@gmail.com>
Date:   Thu Feb 8 10:26:50 2018 -0600

    782: MST MFTF Test Cases

    - Fixed the Title of a Test.
    - Adding scrollToTopOfPage to a Test that had issues with the screen size.

commit 5c2840fdba056f0adc6a69ae95b04d0ac0d02225
Author: John Stennett <john00ivy@gmail.com>
Date:   Wed Feb 7 14:58:47 2018 -0600

    782: MST MFTF Test Cases

    - Removing references to Priority since it was removed from the UI.
    - Replacing "_defaultSource" reference with "TestSource1".
    - Updating the dataType for the Metadata/Data.

commit 7645a6abb93d06a21920010754a2fe0127e78107
Author: John Stennett <john00ivy@gmail.com>
Date:   Tue Feb 6 15:12:39 2018 -0600

    782: MST MFTF Test Cases

    - Adding additional Stock data.
    - Adding UpdateStockDataTest.

commit 0b3fadbd8e6f7cf4840e8368ae11f09af545f1c9
Author: John Stennett <john00ivy@gmail.com>
Date:   Tue Feb 6 14:34:23 2018 -0600

    782: MST MFTF Test Cases

    - Adding Stock meta-data.
    - Renaming AdminManageStockGrid. Adding additional selectors.
    - Renamed a few Test files.
    - Adding Search by Keyword test for Stock.

commit d72ee7b83e5301c60abc714499073908e49e2159
Author: John Stennett <john00ivy@gmail.com>
Date:   Tue Feb 6 13:41:53 2018 -0600

    782: MST MFTF Test Cases

    - Adding the "Assign Sources" slide out section.

commit 98c7d8c300c5ca05fd08b99671d026408387e315
Author: John Stennett <john00ivy@gmail.com>
Date:   Tue Feb 6 13:41:17 2018 -0600

    782: MST MFTF Test Cases

    - Adding Stock Metadata.
    - Adding Stock Page.
    - Adding Stock Test.
    - Renamed a few Sections.
    - Updated Tests to use new names.

commit 9cbe8e7b443bc1988c8314434f9abd4c2dcc7fca
Author: John Stennett <john00ivy@gmail.com>
Date:   Mon Feb 5 15:12:05 2018 -0600

    782: MST MFTF Test Cases

    - Adding additional data to the _defaultSource model.
    - Replacing "TestSource1" with "_defaultSource".
    - Removing test group.
    - Adding "clearGridFilterIfPresent" to clean up before the Test runs.
    - Adding SearchByKeyword test case.

commit 77c14727fb8986775dc64ced9996f7209d835894
Author: John Stennett <john00ivy@gmail.com>
Date:   Mon Feb 5 13:42:32 2018 -0600

    782: MST MFTF Test Cases

    - Initial commit of MFTF Tests.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Moving all Test files from "/app/code/" to "/dev/tests/".

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Adding a new test for New Websites being assigned to the Default Stock automatically.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Adding a new test for Assigning Multiple Sources to a Stock.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Replacing "Acceptance" with "Functional" in the README and composer.json files.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Adding a wait to a test.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Replaced hardcoded values with Data values in 2 tests since a bug was fixed.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Making some of the Source selectors parameterized.
- Adding severities to all of the MSI tests.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Correcting parameterized selectors for Sources.
- Adding new Test.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Replacing selectors and action groups with the correct versions. There were duplicate sections and action groups that were cleaned up.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Renamed Data entities. Updated Tests.
- Added MSI Store/Store View/Website data.
-

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Adding a new test for Create a New Website with a Store and Store View.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Adding a Test for Default Stock cannot be deleted.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Adding new Test for Cannot Delete Stock when Sales Channel is assigned.
- Adding description annotation to a test.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Adding additional Websites for an MSI test.
- Updating Tests to use the new Websites for MSI.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Moving the Test files from "dev/tests/" to "app/code/".

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Skipping a few Tests due to a Bug in MSI. #815 Adding notes to the tests and links for the Bugs.
- Adding a few <waitForPageLoad/> to correct a few timing issues.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Adding a new test for Multiple Websites assigned as Sales Channels.
- Adding a new test for Updating the Product quantity.
- Updating the Product metadata so the API will function correctly.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Adding the final tests for this PR.
- Adding the Deleting Stock from Grid Test.
- Adding the Configurable Product with drop down attribute Test.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Filling in a few missing testCaseId.
- Moving the skip group below the message.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Adding a line at the end of the files.

ivy00johns committed:
Merge branch '2.3-develop' into 833-MSI-MFTF-Test-Cases-2

okolesnyk commented on app/code/Magento/Inventory/Test/Acceptance/Page/AdminManageSourcesPage.xml:
```diff
@@ -0,0 +1,16 @@
+
+<pages xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:noNamespaceSchemaLocation="../../../../../../vendor/magento/magento2-functional-testing-framework/src/Magento/FunctionalTestingFramework/Page/etc/PageObject.xsd">
+    <page name="ManageSources" url="/inventory/source/index/" area="admin" module="MultiSourceInventory">
```
Module must be specified by this template `<vendorName>_<moduleName>` in this case it should be `Magento_Inventory`

okolesnyk replied:
Please update name to `AdminManageSourcePage`

ivy00johns replied:
Updated.

okolesnyk commented on app/code/Magento/Inventory/Test/Acceptance/Page/AdminManageStockPage.xml:
```diff
@@ -0,0 +1,14 @@
+
+<pages xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:noNamespaceSchemaLocation="../../../../../../vendor/magento/magento2-functional-testing-framework/src/Magento/FunctionalTestingFramework/Page/etc/PageObject.xsd">
+    <page name="ManageStock" url="/inventory/stock/index/" area="admin" module="MultiSourceInventory">
```
Module must be specified by this template `<vendorName>_<moduleName>` in this case it should be `Magento_Inventory`

ivy00johns replied:
Updated.

okolesnyk commented on app/code/Magento/Inventory/Test/Acceptance/Section/AdminAssignSourcesSlideOutSection.xml:
```diff
@@ -0,0 +1,20 @@
+        <element name="defaultSourceCheckbox" type="checkbox" selector="#idscheckdefault"/>
+    </section>
+    <section name="AdminAssignSourcesSlideOutGridSection">
+        <element name="checkboxByIndex" type="checkbox" selector=".admin__data-grid-outer-wrap .data-row[data-repeat-index='{{var1}}'] .admin__control-checkbox" parameterized="true"/>
```
I suggest to use xPath here and rely not on index but on Source Code
I verified this selector will work.
`//tr[td[contains(.,"{{code}}")]]//input[@data-action="select-row"]`

ivy00johns replied:
Updated.

okolesnyk commented on app/code/Magento/Inventory/Test/Acceptance/Section/AdminManageSourcesEditFormSection.xml:
```diff
@@ -0,0 +1,45 @@
+        <element name="longitude" type="input" selector="input[name='general[longitude]']"/>
+    </section>
+    <section name="AdminEditSourceContactInfoSection">
+        <element name="opened" type="wysiwyg" selector="div[data-index='contact_info'] .fieldset-wrapper-title[data-state-collapsible='open']"/>
```
On my opinion type should be button, because behavior of this element closer to the button then wysiwyg.
This also applies to other similar elements.

ivy00johns replied:
Updated.

okolesnyk commented on app/code/Magento/Inventory/Test/Acceptance/Section/AdminManageSourcesGridSection.xml:
```diff
@@ -0,0 +1,20 @@
+        <element name="addNewSource" type="button" selector="#add" timeout="30"/>
+    </section>
+    <section name="AdminManageSourcesGridFilterControls">
+        <element name="dropDown" type="button" selector=".admin__data-grid-header .admin__data-grid-filters-current._show"/>
```
Looks like this selector doesn't represent filter button or any drop down. This selector applies to list of applied filters.

ivy00johns replied:
Changed to "block".

okolesnyk commented on app/code/Magento/Inventory/Test/Acceptance/Section/AdminProductSourcesFormSection.xml:
```diff
@@ -0,0 +1,26 @@
+        <element name="assignSources" type="button" selector="button[data-index='assign_sources_button']" timeout="30"/>
+    </section>
+    <section name="AdminProductSourcesGrid">
+        <element name="rowByIndex" type="input" selector="div[data-index='sources'] .data-row[data-repeat-index='{{arg1}}']" parameterized="true"/>
```
I suggest to not rely on row index. We never know when sorting will be changed.
Please rely on source code
Here is a selector:
`//tr[td[div[@data-index="source_code" and contains(.,"{{source_code}}")]]]`

okolesnyk replied:
Please apply same approach for selectors below.

ivy00johns replied:
This section contains a "dragAndDrop" feature and I need to verify that the expected data is in the correct row, so I had to use the Index here.

okolesnyk commented on app/code/Magento/Inventory/Test/Acceptance/Test/AdminAddConfigurableProductWithDropDownAttributeAndDefaultSourceTest.xml:
```diff
@@ -0,0 +1,108 @@
+        <see selector="{{AdminProductFormConfigurationsSection.currentVariationsSkuCells}}" userInput="{{colorProductAttribute3.name}}" stepKey="seeAttributeSku3InField"/>
+        <see selector="td[data-index='quantity_per_source_container']" userInput="100" stepKey="seeQuantityInField1"/>
+        <see selector="td[data-index='quantity_per_source_container']" userInput="100" stepKey="seeQuantityInField2"/>
+        <see selector="td[data-index='quantity_per_source_container']" userInput="100" stepKey="seeQuantityInField3"/>
```
I think this asserts "seeQuantityInField" doing the same verification 3 times. 
Probably you need to update your selector to verify quantity for each product item.

ivy00johns replied:
Good catch! Updated.

okolesnyk commented on app/code/Magento/Inventory/Test/Acceptance/Test/AdminAddMultipleSourcesToProductTest.xml:
```diff
@@ -0,0 +1,68 @@
+        <click selector="{{AdminAssignSourcesSlideOutGridSection.checkboxByIndex('0')}}" stepKey="clickOnCheckbox2"/>
+        <click selector="{{AdminAssignSourcesSlideOutSection.done}}" stepKey="clickOnDone2"/>
+
+        <see selector="{{AdminProductSourcesGrid.rowByIndex('1')}}" userInput="$$createSource1.source[source_code]$$" stepKey="seeSourceIdInGrid1"/>
```
Please try to avoid using Indexes. Please find needed row using source code.

ivy00johns replied:
This section contains a "dragAndDrop" feature and I need to verify that the expected data is in the correct row, so I had to use the Index here.

okolesnyk commented on app/code/Magento/Inventory/Test/Acceptance/Test/AdminAssignMultipleSourcesToStockTest.xml:
```diff
@@ -0,0 +1,105 @@
+        <click selector=".page-main-actions button.action-primary" stepKey="clickOnDone4"/>
+        <waitForPageLoad time="30" stepKey="waitForPageLoad6"/>
+
+        <see selector="{{AdminProductSourcesGrid.rowByIndex('0')}}" userInput="$$createSource1.source[source_code]$$" stepKey="seeSource1CodeInGrid1"/>
```
Please avoid using Indexes. Please use source code to find needed row.

ivy00johns replied:
This section contains a "dragAndDrop" feature and I need to verify that the expected data is in the correct row, so I had to use the Index here.

okolesnyk reviewed (changes requested):
1. Please update all paths to XSD files
2. Please verify that modularity of tests are appropriate and in case some module will be disabled tests will not fail.
3. Please do not rely on indexes in selectors. Any other module can extend/update your test and indexes can be different.
4. Paths in composer.json files are wrong because of tests location. Please update them according test location in `app/code`.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Fixing a few tests that won't fun if there isn't an additional source listed.

ivy00johns committed:
Merge remote-tracking branch 'origin/833-MSI-MFTF-Test-Cases-2' into 833-MSI-MFTF-Test-Cases-2

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Adding additional elements on the Add Source page.
- Moving hardcoded selectors to Section Objects.
- Updating Tests to use the Section Objects instead of hardcoded selectors.
- Removing extra <waitForPageLoad> actions for elements that have "timeout" values attacted to the <element>.
- Removing unnecessary <conditionalClicks> for the "Reset Filters" action that is included in the "AdminGridFilterSearchResultsByInput" action group.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Removing redundant "conditionalClick" actions.
- Adjusting "time" values for a few actions.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Updating the XSD file paths for all files.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Adding Sections to the correct Pages for Sources/Stocks.
- Updating the Page Names and Module Names.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Adding new <element>'s for the Sources and Stocks pages.
- Adjusting the <element> "type"'s for a few elements.
- Updating a few tests to use "rowByCode" instead of "currentVariationsQuantity".
- Updating a few tests to use "checkboxByCode" instead of "checkboxByIndex".
- Updating the Page Names for the Stock and Source pages.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Corrected a data reference in Test.

ivy00johns commented:
This PR depends on this Magento2CE CR to Magento Core: 
* https://fisheye.corp.magento.com/cru/CR-MAGETWO-38862

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Replacing "block" with "textarea" since 'block' was removed from the XSD.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Blacklisting the XML files from the Integrity tests until we support URN's in MFTF.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Blacklisting the XML files from the Integrity tests until we support URN's in MFTF.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Correcting paths in the composer.json files.

ivy00johns committed:
833: MSI MFTF Test Cases 2

- Excluding the SchemeTest until MFTF supports URNs.

ivy00johns commented:
Merged.
